src/main/java/de/bonndan/nivio/input/ItemDescriptionFormatFactory.java

```diff
@@ -1,44 +1,58 @@
 
+@Service
 public class ItemDescriptionFormatFactory {
 
+    private final Map<ItemDescriptionFactory, List<String>> factoryListMap = new ConcurrentHashMap<>();
+
+    public static ItemDescriptionFormatFactory with(ItemDescriptionFactory factory) {
+        return new ItemDescriptionFormatFactory(new ArrayList<>(Arrays.asList(factory)));
+    }
+
+    public ItemDescriptionFormatFactory(List<ItemDescriptionFactory> factories) {
+        factories.forEach(itemDescriptionFactory -> factoryListMap.put(itemDescriptionFactory, itemDescriptionFactory.getFormats()));
+    }
+
     /**
      * Returns the proper factory to generate/parse service descriptions based on the input format.
      *
-     * @param reference the reference pointing at a file or url
+     * @param reference            the reference pointing at a file or url
      * @param landscapeDescription landscape, may contain a base url
      * @return the factory
      */
-    public static ItemDescriptionFactory getFactory(SourceReference reference, LandscapeDescription landscapeDescription) {
-
-        URL baseUrl = URLHelper.getParentPath(landscapeDescription.getSource());
-
-        FileFetcher fetcher = new FileFetcher(new HttpService());
-
-        if (reference == null || reference.getFormat() == null)
-            return new ItemDescriptionFactoryNivio(fetcher, baseUrl);
-
-        //TODO use SourceFormat.from or similar
-        switch (reference.getFormat()) {
-            case DOCKER_COMPOSE2:
-                return new ItemDescriptionFactoryCompose2(new FileFetcher(new HttpService()), baseUrl);
-            case KUBERNETES:
-                return new ItemDescriptionFactoryKubernetes(reference);
-            case RANCHER1_PROMETHEUS:
-                return new ItemDescriptionFactoryRancher1(baseUrl);
+    public ItemDescriptionFactory getFactory(SourceReference reference, LandscapeDescription landscapeDescription) {
+
+        List<ItemDescriptionFactory> factories = new ArrayList<>();
+        factoryListMap.entrySet().stream()
+                .filter(entry -> entry.getValue().stream().map(s -> {
+                    if (StringUtils.isEmpty(s))
+                        return "";
+                    return s.toLowerCase();
+                }).anyMatch(s -> s.equals(reference.getFormat()) || (StringUtils.isEmpty(s) && StringUtils.isEmpty(reference.getFormat()))))
+                .forEach(entry -> factories.add(entry.getKey()));
+
+        if (factories.isEmpty()) {
+            List<String> knownFormats = new ArrayList<>();
+            factoryListMap.values().forEach(knownFormats::addAll);
+            String msg = "Unknown source reference format: '" + reference.getFormat() + "', known formats are: "
+                    + StringUtils.collectionToDelimitedString(knownFormats, ", ");
+            throw new ProcessingException(landscapeDescription, msg);
         }
 
-        return new ItemDescriptionFactoryNivio(fetcher, baseUrl);
+        //last one wins
```

Comment:
Ich bin jetzt noch nicht so tief drin im Code: Aber wie kann es denn passieren, dass mehrere Factories registriert sind?

Reply:
Spring registers all @Service annotated ItemDescriptionFactory implementations here:

```java
public ItemDescriptionFormatFactory(List<ItemDescriptionFactory> factories) {
        ...
    }
```

My idea was to map the factories to formats the are able to process, like "rancher1" etc. Nivio is meant to have a plugin structure #14  and hence you could add your own plugins/formats or overwrite existing ones. Once #14 is implemented the correct order has to be guaranteed, of course.